packages/ui/react-ui-editor/src/components/Toolbar/Toolbar.stories.tsx

```diff
@@ -4,61 +4,73 @@
 
 import '@dxosTheme';
 
-import React, { type FC, useMemo, useState } from 'react';
+import React, { type FC, useState } from 'react';
 
 import { TextObject } from '@dxos/echo-schema';
 import { PublicKey } from '@dxos/keys';
 import { faker } from '@dxos/random';
-import { Tooltip } from '@dxos/react-ui';
-import { mx, textBlockWidth } from '@dxos/react-ui-theme';
+import { Tooltip, useThemeContext } from '@dxos/react-ui';
+import { textBlockWidth } from '@dxos/react-ui-theme';
 import { withTheme } from '@dxos/storybook-utils';
 
 import { Toolbar } from './Toolbar';
 import {
   decorateMarkdown,
   comments,
-  formatting,
-  image,
+  formattingKeymap,
   table,
   useComments,
   useFormattingState,
+  createMarkdownExtensions,
+  image,
 } from '../../extensions';
-import { type Comment, useActionHandler, useEditorView, useTextModel } from '../../hooks';
+import {
+  type Comment,
+  createBasicExtensions,
+  createDataExtensions,
+  createThemeExtensions,
+  useActionHandler,
+  useDocAccessor,
+  useTextEditor,
+} from '../../hooks';
 import translations from '../../translations';
-import { MarkdownEditor } from '../TextEditor';
 
 faker.seed(101);
 
-const Story: FC<{ id?: string; content: string }> = ({ id = 'test', content }) => {
+const Story: FC<{ content: string }> = ({ content }) => {
+  const { themeMode } = useThemeContext();
   const [item] = useState({ text: new TextObject(content) });
-  const [_comments, setComments] = useState<Comment[]>([]);
-  const [editorRef, viewInvalidated] = useEditorView(id);
-  const model = useTextModel({ text: item.text });
+  const { id, doc, accessor } = useDocAccessor(item.text);
   const [formattingState, formattingObserver] = useFormattingState();
-  const extensions = useMemo(
-    () => [
-      decorateMarkdown(),
-      comments({
-        onCreate: ({ cursor }) => {
-          const id = PublicKey.random().toHex();
-          setComments((comments) => [...comments, { id, cursor }]);
-          return id;
-        },
-      }),
-      formatting(),
-      image(),
-      table(),
-      formattingObserver,
-    ],
-    [],
+  const { parentRef, view } = useTextEditor(
+    {
+      doc,
+      extensions: [
+        createBasicExtensions(),
+        createMarkdownExtensions({ themeMode }),
+        createThemeExtensions({ themeMode, slots: { editor: { className: 'p-2' } } }),
+        createDataExtensions({ id, text: accessor }),
+        comments({
+          onCreate: ({ cursor }) => {
+            const id = PublicKey.random().toHex();
+            setComments((comments) => [...comments, { id, cursor }]);
+            return id;
+          },
+        }),
+        decorateMarkdown(),
+        formattingKeymap(),
+        image(),
+        table(),
+        formattingObserver,
+      ],
```

Comment:
Memoize extensions? It’s just a story, but folks will copypaste from stories, so probably worth setting an example for best practices.